spec/lib/activerecord_spec.rb

```diff
@@ -100,6 +100,95 @@
+      record.remove_data_entries(sources: data_source.name)
+
+      expect { record.forge! }.not_to change { record.reload.name }
+    end
+  end
+
+  describe '#forge' do
+    let(:record) { ActiveRecordMock.create(name: "Post", active: true) }
+    let(:data_source) { LaForge::DataSource.find_or_create_by(name: "bbc", priority: 1) }
+    let(:data_entry) { LaForge::DataEntry.new(record: record, source_id: data_source.id) }
+
+    let(:prioritized_data_source) { LaForge::DataSource.find_or_create_by(name: "guardian", priority: 2) }
+    let(:prioritized_data_entry) { LaForge::DataEntry.new(record: record, source_id: prioritized_data_source.id) }
+
+    it 'sets the record attributes from the data entries' do
+      data_entry.update_attributes(attribute_name: "name", value: "Article")
+      expect { record.forge }.to change { record.name }.from("Post").to("Article")
+    end
+
+    it 'does not change the database record' do
+      data_entry.update_attributes(attribute_name: "name", value: "Article")
+      expect { record.forge }.not_to change { record.reload.name }
+    end
+
+    it 'ignores data_entries with an invalid attribute name' do
+      data_entry.update_attributes(attribute_name: "name_invalid")
+      expect { record.forge }.not_to change { record.changes }
+    end
+
+    it 'merges data_entries from multiple sources' do
+      data_entry.update_attributes(attribute_name: "name", value: "Article")
+      prioritized_data_entry.update_attributes(attribute_name: "active", value: false)
+
+      expect { record.forge }.to change { record.changes }.to eq({"name"=>["Post", "Article"], "active"=>[true, false]})
+    end
+
+    it 'prioritizes based on the data_entry priority' do
+      data_entry.update_attributes(attribute_name: "name", value: "Higher Priority", priority: 5)
+      prioritized_data_entry.update_attributes(attribute_name: "name", value: "Lower Priority", priority: 3)
+
+      expect { record.forge }.to change { record.name }.from("Post").to("Higher Priority")
+    end
+
+    it 'prioritizes based on the data_source priority when priority is not set on the data_entry' do
+      data_entry.update_attributes(attribute_name: "name", value: "Lower Priority")
+      prioritized_data_entry.update_attributes(attribute_name: "name", value: "Higher Priority")
+
+      expect { record.forge }.to change { record.name }.from("Post").to("Higher Priority")
+    end
+  end
+
+  describe '#forge_attributes' do
+    let(:data_source) { LaForge::DataSource.find_or_create_by(name: "bbc", priority: 1) }
+    let(:record) { ActiveRecordMock.create(name: "Post", active: true) }
+    let!(:data_entry) { LaForge::DataEntry.create!(record: record, attribute_name: "name", value: "Article", source_id: data_source.id) }
+
+    it 'returns a hash of the attributes generated the data entries' do
+      expect(record.forge_attributes).to eq({"name"=> 'Article'})
+    end
+
+    it 'removes data_entries for invalid columns' do
```

Comment:
Let's be explicit here as to what we mean by "invalid columns", again referencing the setter.